src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509EnhancedKeyUsageExtension.cs

```diff
@@ -42,7 +42,10 @@ public OidCollection EnhancedKeyUsages
                     X509Pal.Instance.DecodeX509EnhancedKeyUsageExtension(RawData, out _enhancedKeyUsages);
                     _decoded = true;
                 }
-                return _enhancedKeyUsages;
+                OidCollection oids = new OidCollection();
+                foreach (Oid oid in _enhancedKeyUsages)
+                    oids.Add(oid);
+                return oids;
```

Comment:
What's the benefit of returning a new collection each time, other than exact consistency with how netfx does it?  Properties like this aren't generally expected to return a new collection each time, especially since it means code that expects to be able to do:
```C#
c.EnhancedKeyUsages.Add(...);
```
will think it's being added but it's actually a nop.

Reply:
NetFX does exactly this. You should not be able to modify the object by calling `c.EnhancedKeyUsages.Add(...)`, it would result in inconsistent internal state.

Reply:
In ideal world there would a read-only version of `OidCollection` that would throw an exception, but that would be a big API change. This ensures that the object cannot be corrupted by the caller.

I am fine if you decide it's intentional change to the API behaviour and drop this PR, but I filed an issue for it beforehand and it was marked as 'bug'.

Reply:
> NetFX does exactly this

I understand that.

> you should not be able to modify the object by calling c.EnhancedKeyUsages.Add(...)

That may be, but this is a very unexpected pattern, *and* it makes it more expensive to access this data, requiring multiple additional allocations every time the property is accessed.  If the goal is to avoid code calling Add from successfully mutating the object, there are other ways to accomplish that.  I'll let @bartonjs decide what to do here, this just seems wrong to me.

Reply:
In principle I agree with you. Do you have any better implementation in mind? Adding a read-only mode for OidCollection and exposing it using internal properties/methods (eg. OidCollection.AsReadOnly)?

Reply:
@stephentoub Of the places in corefx that touch this property, it gets used in a foreach; so pretty much the object gets created zero or one times... except in the Mono version of the new test in this PR (and so now a second place).

Scanning through consumption on github I find mostly copies of corefx and the Mono tests; but the actual callers I found also touch it precisely once.

There are two observable differences between the current corefx implementation and the netfx implementation:
1) ReferenceEquals(eku.EnhancedKeyUsages, eku.EnhancedKeyUsages) differs between them
2) How mutation is observed (second property hit vs local)

On the one hand, the utilization of this property is so low (and seem to not notice either of those points) that introducing a behavioral change probably doesn't matter.  On the other hand, it's also not hot code for anyone, so it's not like we would have done the inverse of this change to reduce the creation of the new collection had it been ported more faithfully.

If the rest of the certificate was "pure" I'd say this is an easy yes; but apparently the cert Extensions collection doesn't guard against mutation.  And these Oid values are shared between the property-returned collection and the held collection, and Oid is surprisingly mutable.  So it's just a question of what our compatability goals are right now for corefx as a whole.

Reply:
> so pretty much the object gets created zero or one times

It's the extra one time that I'm pushing on. You know more about this than I do, so if you say it's not a big deal, fine.  But I see that from SocketsHttpHandler when setting up an SSL connection with client certs:
https://source.dot.net/#System.Net.Http/System/Net/Http/HttpClientHandler.Windows.cs,346
for every cert we check whether it's a contender:
https://source.dot.net/#System.Net.Http/CertificateHelper.cs,44
and for each of that cert's extensions we check if it's valid:
https://source.dot.net/#System.Net.Http/CertificateHelper.cs,57
and checking if it's valid accesses EnhancedKeyUsages:
https://source.dot.net/#System.Net.Http/CertificateHelper.cs,72
which with this PR will incur an extra `OidCollection` allocation, `List<Oid>` allocation in the `OidCollection`, and `Oid[]` allocation in the `List<Oid>`.  So three additional allocations per extension per certificate.  I'm simply suggesting that overhead would be nice to avoid.

Reply:
Given all the arguments I am leaning towards dropping this PR. I can submit a different PR to implement OidCollection.AsReadOnly, which would reuse the internal list, but disable the Add method (throw an exception even?). The behaviour would still be observable, but it should not break existing code as long as no one depends on the collection being a copy that they save for their own use.

Reply:
The one use case I care about here is perf with SocketsHttpHandler.  If this change doesn't measurably and negatively affect that (maybe it's a drop in the bucket on the costs of setting up a connection), then I'm fine with matching the netfx behavior here.

Reply:
...or even better I can implement it as copy-on-write, where OidCollection.Add would do the list copy. That should not be observable at all and the performance impact should be minimal.

Reply:
@filipnavara Since OidCollection is in a different assembly that'd require new public API (assuming I understood the intent correctly).